### ISSUE TITLE
Consider NamedFiledPuns declarations

### ISSUE THREAD

**Issue by roman** [lspitzner/brittany#77]
In some parts of my project I'm using NamedFieldPuns GHC extension, and on some occasions, I'm getting a whole lot of symbols out of a record, e.g.

```haskell
runSupervisorLoop :: SupervisorEnv -> IO ()
runSupervisorLoop env@SupervisorEnv { supervisorId, supervisorName, supervisorStatusVar, supervisorQueue, notifyEvent }
```

Ideally, I would like Brittany to do a newline per field name when the list of symbols is bigger than the max column setting.

**Comment by lspitzner** [comment]
Right, i was lazy when implementing the pattern-match layouter and only did the single-line version.

> Ideally, I would like Brittany to do a newline per field name when the list of symbols is bigger than the max column setting.

Agreed. The first step in implementing this is a more detailed description of the expected layout for such cases. What is the optimal layout for:

~~~~.hs
runSupervisorLoop env@SupervisorEnv { supervisorId, supervisorName, supervisorStatusVar, supervisorQueue, notifyEvent } = large multiline body
~~~~

?

(And really, this is still a simple case - what if there are multiple arguments with record accesses? Now add a guard, or multiple guards. Now add Viewpatterns in those guards which contain records with NamedFieldPuns again. To properly implement multiline pattern-matches, all this needs to be considered. Not to say that all these questions need to be answered up front - we can simply resort to the (current) only-allow-single-line-layout. Just a bit of background why I shied away from this whole thing..)

I simply don't encounter very long patterns very often, and as a consequence i have not much of an idea of how to layout them consistently over multiple lines. I can help with getting things implemented, but for this requirement analysis/design I am no more qualified than any other haskell users :p

**Comment by tfausak** [comment]
I have not found a good way to lay out long function definitions. It's a weird corner of Haskell's syntax. Usually when I run into a situation where my function definition is too long, I change something about the function. For example:

``` hs
runSupervisorLoop env = do
  let SupervisorEnv { supervisorId, supervisorName, supervisorStatusVar, supervisorQueue, notifyEvent } = env
  ...
```

Obviously that just passes the buck and doesn't really answer how to format this 😄

**Comment by roman** [comment]
Hey @lspitzner,

First, thanks for answering on a Sunday afternoon, wasn't expecting that at all.

In regards to your comment:

>  Now add Viewpatterns in those guards which contain records with NamedFieldPuns again. To properly implement multiline pattern-matches, all this needs to be considered. Not to say that all these questions need to be answered up front - we can simply resort to the (current) only-allow-single-line-layout for parts. Just a bit of background why I shied away from this whole thing..)

![escalated quickly](http://i0.kym-cdn.com/photos/images/facebook/000/353/279/e31.jpg)

Yikes, yeah, Haskell syntax is _simple_, right?

With all that in place, I wonder what would be considered a good middle ground, also, what is the next bit of that whole list to tackle. If you feel like it would be quicker, you can point me to the spot where this bit is being handled and offer some suggestions on how to move forward. I've already read the doc folder, and everything seems to make sense, the code is significant, and I don't have a good idea of what module does what at this point, so I appreciate any pointers.

Cheers.

**Comment by lspitzner** [comment]
@roman i don't think there is a good middle ground, but that should not stop us from starting the change from "consistently ignore the column limit in pattern matches" to "never ignore the column limit". In the worst case, we can hide the new, partially completed behaviour behind a config setting for those that don't mind somewhat reduced consistency.

So really, we can start with the example you provided initially. Brittany can express "if there are any guards, or more than one argument, use the current behaviour, but otherwise, also allow this multiline-layout" just fine.

Another interesting case is: No guards, but just a long list of arguments. E.g.

~~~~.hs
func aaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbb cccccccccccc ddddddddddddddddddd = do {..}
~~~~

The code responsible for layouting patterns is [in this module](https://github.com/lspitzner/brittany/blob/63c04e0b661276416992fbe4d0188f4d7b8e2707/src/Language/Haskell/Brittany/Internal/Layouters/Pattern.hs#L27). I'll be honest and admit that i'll have to reconsider how it works, too (why exactly it returns a `Seq BriDocNumbered` instead the usual `BriDocNumbered` etc.). I have added some comments in 0036dbf41086030083ba9aec3f0165c06fe62f30.

(reminder to please base changes on the `dev` branch, even when `dev`/`master` are mostly in sync right now)

I'd really like to see some example multiline layouts/testcases before any code changes, though.

**Comment by lspitzner** [comment]
(btw the more general goal was already open as #14, but I don't mind keeping both issues open for some time.)

**Comment by cblp** [comment]
there is a typo in the title of the issue: NamedFi[le]dPuns

**Comment by cblp** [comment]
currently, I see ugly `New {newText , newStart , newEnd }`
